benchmark/Cpp/Savina/src/concurrency/CigaretteSmoker.lf

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright (C) 2020 TU Dresden
+ * 
+ * This benchmark "solves" the Cigarette Smoker Problem similarly to the
+ * original Savina implementation in Akka. However, it is important to note
+ * first, that the Savina implementation does not actually model the resources
+ * (paper, matches, tobacco) and thus does not provide a complete solution. It
+ * only uses a pattern similar to a common solution of the Cigaratte Smokers
+ * Problem. This solution uses an arbitrator that tells the smokers when they
+ * can grab the resources and smoke. Since no resources are modelled in Savina,
+ * the arbitrator actually just randomly selects a smoker and tells it to smoke.
+ *
+ * The process of smoking is implemented in the Savina suite by calling Java
+ * `Math.random()` a few times in a loop. Converting this "workload" to another
+ * language is non-trivial, as likely no similar function with the precise same
+ * implementation exists. For the C++ port here, we use `rand()` from stdlib.
+ * This is a significant restriction of this benchmark and limits the
+ * comparability of results obtained from Akka and from LF C++.
+ *
+ * Since actors operate asynchronously, several smokers can actually smoke in
+ * parallel. In theory, this allows for a fast execution, as the workload of
+ * each smoker can be processed in parallel by many smokers. Achieving this in
+ * LF, however, is non-trivial. In LF, all the communication is synchronous and
+ * instantaneous. When the arbitrator tells a smoker to smoke, the smoker will
+ * start smoking at the same tag and the execution will wait until this smoker
+ * finished. Only then the runtime will forward logical time to the next tag and
+ * the arbitrator will tell the next smoker to smoke. Obviously, there is no
+ * parallelism in this solution.
```

Comment:
The point of this challenge problem to achieve coordinated mutual exclusion (i.e., without deadlock) between concurrent processes. If parallel smoking is desired, the coordinator can put more than one ingredient on the table and thus signal multiple smokers to smoke simultaneously. This is easy to do in LF...

Reply:
I didn't describe the problem very well (I actually confused myself while reading it again...). I adjusted the text and hope my point is clearer now.

The cigarette smoker problem is a little more complex. The smokers implement two operations: making a cigarette and smoking. They first need to make a cigarette before they can smoke, and for making a cigarette they need access to the ingredients (resources). Since access to the resources is limited, only a single smoker can make a cigarette at any point in time. However, all the smokers that already have a cigarette can smoke whenever they like and smoking can take arbitrarily long. It is this parallelism between the cigarette smoking process and the arbitration and cigarette making that is hard to achieve in LF. In order to make a new cigarette, we need to advance time (because simultaneous access to the resources is not allowed, and also to avoid cycles), but due to the barrier synchronization we need to wait for all smoking reactions to finish before we can advance time.

Reply:
It seems that it would behove us to explicitly model the timing in the problem.  The performance bottleneck you cite is because smoking takes zero (logical) time.

Reply:
I don't understand how adding a logical delay for smoking would increase the performance here. The problem arises from the barrier synchronization in our runtime. This would of course look differently in a federated execution, but in the non-federeated runtime, we are forced to wait until all reactions complete before we can advance time and give the ingredients to the next smoker. I think it would be worthwhile though to think about ways to relax the barrier synchronization here.